### DISCUSSION TITLE
Change approach used for memory buffers

### DISCUSSION THREAD

**Discussion by david3a** [vsf-tv/gccg-api#30] [General]
This is targetted at better matching memory allocation strategy such that sending is page aligned for each "grain"

On the TX side this would change the setup parameters from

* @param tx_buffer_size_bytes The size in bytes of a memory region for holding transmit payload data. A pointer to the
*                             buffer is returned in ret_tx_buffer_ptr. The application manages how the buffer is
*                             partitioned and used.

* @param ret_tx_buffer_ptr Pointer where to write the returned start of allocated transmit payload buffer. Size is specified
*                          using tx_buffer_size_bytes.

to:

* @param tx_buffer_size_bytes The size in bytes of a memory region for holding a single transmit payload data. 
*
* @param tx_buffer_count Positive integer count of buffers requested by the application for sending
* 
* @param ret_tx_buffer_ptrs Pointer to an array of pointers to receive allocated transmit payload buffers of the size requested.
* 

This will allow all buffers to be page aligned at the start which facilitates interface with several potential subsystems, the provider may allocate more memory to ensure the buffer is a multiple of page size if required.

This will include providers that work with a single large buffer or those for which multiple non-contiguous buffers would be preferred   

For receive I propose only that we change

 * @param rx_buffer_size_bytes The size in bytes of a memory region for holding received payload data.

to

 * @param rx_buffer_size_bytes The size in bytes of a memory region for holding received a single payload for a "grain".

this allows the provider to manage the memory used effectively and the allocation of buffers to be received into such that we avoid overruns and wrap-around issues

I'll prepare a pull request along these lines

The only other potential change I'd suggest would be to add a function to allow a TX application to request the buffer to use for the next TX cycle, again this allows the provider to determine its best strategy in the allocation of buffers, is this approach was taken then we would not need to return an array of buffers in the setup function

/**
 * Request a buffer for a single TX grain. This API is thread-safe.
 *
 * @param handle Connection handle returned by the GccgTxConnectionCreate() API functions.
 *
* @param buffer Receives buffer to be used for the next call to GccgTxPayload
 * @return A value from the GccgReturnStatus enumeration.
 */

GCCG_INTERFACE GccgReturnStatus GccgRequestTxBuffer(GccgConnectionHandle handle,
                                              GccgMediaElement *buffer );

If used for multiple flows (components) on the same connection will need a further tweak, we would not expect to support that behaviour as our processing is always component based

**Comment by pac-work** [comment]
Hello, while designing our internal communication system, we have landed on very similar parametrization of the interface, so I can only second this proposal.

Just a small remark – careful with the use of the word “grain” in the comments. For example, when doing TX, the standard defines transferring multiple grains (audio + video + anc.) in one payload (buffer being transferred). At least that is how I understand it – all elements (`GccgMediaElement`) pointed to from the `media_array` are inside 1 buffer. Or am I wrong on this?

**Reply by david3a** [reply]
I think you are correct on the potential to send multiple components in one transfer, I'd prefer to see a separation so that the transport is able to manage buffer sizes suitably.
